diff --git a/bin/storm b/bin/storm
index bb6e40a..1df1bb0 100755
--- a/bin/storm
+++ b/bin/storm
@@ -42,7 +42,18 @@ else:
 
 STORM_DIR = "/".join(os.path.realpath( __file__ ).split("/")[:-2])
 USER_CONF_DIR = os.path.expanduser("~/.storm")
-CLUSTER_CONF_DIR = STORM_DIR + "/conf"
+STORM_CONF_DIR = os.getenv('STORM_CONF_DIR', None)
+
+if STORM_CONF_DIR == None:
+    CLUSTER_CONF_DIR = STORM_DIR + "/conf"
+else:
+    CLUSTER_CONF_DIR = STORM_CONF_DIR
+
+STORM_LOG_DIR = os.getenv('STORM_LOG_DIR', None) 
+
+if (STORM_LOG_DIR == None):
+    STORM_LOG_DIR = STORM_DIR+"/logs"
+
 if (not os.path.isfile(USER_CONF_DIR + "/storm.yaml")):
     USER_CONF_DIR = CLUSTER_CONF_DIR
 CONFIG_OPTS = []
@@ -133,6 +144,7 @@ def exec_storm_class(klass, jvmtype="-server", jvmopts=[], extrajars=[], args=[]
     all_args = [
         "java", jvmtype, get_config_opts(),
         "-Dstorm.home=" + STORM_DIR, 
+        "-Dstorm.log.dir=" + STORM_LOG_DIR,
         "-Djava.library.path=" + confvalue("java.library.path", extrajars),
         "-Dstorm.conf.file=" + CONFFILE,
         "-cp", get_classpath(extrajars),
diff --git a/logback/cluster.xml b/logback/cluster.xml
index a761c12..35ea63f 100644
--- a/logback/cluster.xml
+++ b/logback/cluster.xml
@@ -18,9 +18,9 @@
 
 <configuration scan="true" scanPeriod="60 seconds">
  <appender name="A1" class="ch.qos.logback.core.rolling.RollingFileAppender">
-    <file>${storm.home}/logs/${logfile.name}</file>
+    <file>${storm.log.dir}/${logfile.name}</file>
     <rollingPolicy class="ch.qos.logback.core.rolling.FixedWindowRollingPolicy">
-      <fileNamePattern>${storm.home}/logs/${logfile.name}.%i</fileNamePattern>
+      <fileNamePattern>${storm.log.dir}/${logfile.name}.%i</fileNamePattern>
       <minIndex>1</minIndex>
       <maxIndex>9</maxIndex>
     </rollingPolicy>
@@ -35,9 +35,9 @@
  </appender>
 
  <appender name="ACCESS" class="ch.qos.logback.core.rolling.RollingFileAppender">
-    <file>${storm.home}/logs/access.log</file>
+    <file>${storm.log.dir}/access.log</file>
     <rollingPolicy class="ch.qos.logback.core.rolling.FixedWindowRollingPolicy">
-      <fileNamePattern>${storm.home}/logs/access.log.%i</fileNamePattern>
+      <fileNamePattern>${storm.log.dir}/access.log.%i</fileNamePattern>
       <minIndex>1</minIndex>
       <maxIndex>9</maxIndex>
     </rollingPolicy>
@@ -52,7 +52,7 @@
   </appender>
 
   <appender name="METRICS" class="ch.qos.logback.core.rolling.RollingFileAppender">
-    <file>${storm.home}/logs/metrics.log</file>
+    <file>${storm.log.dir}/metrics.log</file>
     <rollingPolicy class="ch.qos.logback.core.rolling.FixedWindowRollingPolicy">
       <fileNamePattern>${storm.home}/logs/metrics.log.%i</fileNamePattern>
       <minIndex>1</minIndex>
diff --git a/storm-core/src/clj/backtype/storm/daemon/supervisor.clj b/storm-core/src/clj/backtype/storm/daemon/supervisor.clj
index 2471626..e0f74da 100644
--- a/storm-core/src/clj/backtype/storm/daemon/supervisor.clj
+++ b/storm-core/src/clj/backtype/storm/daemon/supervisor.clj
@@ -448,6 +448,7 @@
     :distributed [supervisor storm-id port worker-id]
     (let [conf (:conf supervisor)
           storm-home (System/getProperty "storm.home")
+          storm-log-dir (System/getProperty "storm.log.dir")
           stormroot (supervisor-stormdist-root conf storm-id)
           stormjar (supervisor-stormjar-path stormroot)
           storm-conf (read-supervisor-storm-conf conf storm-id)
@@ -464,6 +465,7 @@
                     [(str "-Djava.library.path=" (conf JAVA-LIBRARY-PATH))
                      (str "-Dlogfile.name=" logfilename)
                      (str "-Dstorm.home=" storm-home)
+                     (str "-Dstorm.log.dir=" storm-log-dir)
                      (str "-Dlogback.configurationFile=" storm-home "/logback/cluster.xml")
                      (str "-Dstorm.id=" storm-id)
                      (str "-Dworker.id=" worker-id)
