diff -pruN spark-1.2.0-bin-hadoop2.4-orig/bin/spark-submit spark-1.2.0-bin-hadoop2.4/bin/spark-submit
--- spark-1.2.0-bin-hadoop2.4-orig/bin/spark-submit	2015-02-13 13:53:46.809606000 -0800
+++ spark-1.2.0-bin-hadoop2.4/bin/spark-submit	2014-12-10 03:00:25.000000000 -0800
@@ -42,12 +42,7 @@ while (($#)); do
   shift
 done
 
-if [ "${SPARK_CONF_DIR}X" == "X" ]
-then
-	DEFAULT_PROPERTIES_FILE="$SPARK_HOME/conf/spark-defaults.conf"
-else
-	DEFAULT_PROPERTIES_FILE="$SPARK_CONF_DIR/spark-defaults.conf"
-fi
+DEFAULT_PROPERTIES_FILE="$SPARK_HOME/conf/spark-defaults.conf"
 export SPARK_SUBMIT_DEPLOY_MODE=${SPARK_SUBMIT_DEPLOY_MODE:-"client"}
 export SPARK_SUBMIT_PROPERTIES_FILE=${SPARK_SUBMIT_PROPERTIES_FILE:-"$DEFAULT_PROPERTIES_FILE"}
 
