diff -pruN spark-1.2.0-bin-hadoop2.4-orig/bin/spark-submit spark-1.2.0-bin-hadoop2.4/bin/spark-submit
--- spark-1.2.0-bin-hadoop2.4-orig/bin/spark-submit	2015-02-13 14:13:18.239933000 -0800
+++ spark-1.2.0-bin-hadoop2.4/bin/spark-submit	2015-02-13 14:13:11.010172000 -0800
@@ -42,7 +42,12 @@ while (($#)); do
   shift
 done
 
-DEFAULT_PROPERTIES_FILE="$SPARK_HOME/conf/spark-defaults.conf"
+if [ "${SPARK_CONF_DIR}X" == "X" ]
+then
+	DEFAULT_PROPERTIES_FILE="$SPARK_HOME/conf/spark-defaults.conf"
+else
+	DEFAULT_PROPERTIES_FILE="$SPARK_CONF_DIR/spark-defaults.conf"
+fi
 export SPARK_SUBMIT_DEPLOY_MODE=${SPARK_SUBMIT_DEPLOY_MODE:-"client"}
 export SPARK_SUBMIT_PROPERTIES_FILE=${SPARK_SUBMIT_PROPERTIES_FILE:-"$DEFAULT_PROPERTIES_FILE"}
 
