diff -pruN spark-1.0.0-bin-hadoop2-orig/bin/compute-classpath.sh spark-1.0.0-bin-hadoop2/bin/compute-classpath.sh
--- spark-1.0.0-bin-hadoop2-orig/bin/compute-classpath.sh	2014-05-26 00:33:30.000000000 -0700
+++ spark-1.0.0-bin-hadoop2/bin/compute-classpath.sh	2014-06-03 15:20:06.125076000 -0700
@@ -28,7 +28,11 @@ FWDIR="$(cd `dirname $0`/..; pwd)"
 . $FWDIR/bin/load-spark-env.sh
 
 # Build up classpath
-CLASSPATH="$SPARK_CLASSPATH:$SPARK_SUBMIT_CLASSPATH:$FWDIR/conf"
+if [ "$SPARK_CONF_DIRX" != "X" ]; then
+  CLASSPATH="$SPARK_CLASSPATH:$SPARK_SUBMIT_CLASSPATH:$SPARK_CONF_DIR"
+else
+  CLASSPATH="$SPARK_CLASSPATH:$SPARK_SUBMIT_CLASSPATH:$FWDIR/conf"
+fi
 
 ASSEMBLY_DIR="$FWDIR/assembly/target/scala-$SCALA_VERSION"
 
