diff -pruN hbase-0.98.17-hadoop2-alternate-ssh/bin/hbase-config.sh hbase-0.98.17-hadoop2/bin/hbase-config.sh
--- hbase-0.98.17-hadoop2-alternate-ssh/bin/hbase-config.sh	2016-01-15 19:27:54.000000000 -0800
+++ hbase-0.98.17-hadoop2/bin/hbase-config.sh	2016-08-04 13:19:57.607497000 -0700
@@ -46,6 +46,8 @@ if [ -z "$HBASE_HOME" ]; then
   export HBASE_HOME=`dirname "$this"`/..
 fi
 
+myhostname=`hostname`
+
 #check to see if the conf dir or hbase home are given as an optional arguments
 while [ $# -gt 1 ]
 do
@@ -53,7 +55,20 @@ do
   then
     shift
     confdir=$1
+    if echo $confdir | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+       orig_confdir="$1"
+       confdir=$(echo "$confdir" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+    if [ ! -d "$confdir" ]; then
+      echo "Error: Cannot find configuration directory: $confdir"
+      exit
+    fi
     shift
+    if [ "${orig_confdir}X" != "X" ]
+    then
+      orig_hbaseconfdir=$orig_confdir
+    fi 
     HBASE_CONF_DIR=$confdir
   elif [ "--hosts" = "$1" ]
   then
@@ -69,6 +84,25 @@ done
  
 # Allow alternate hbase conf dir location.
 HBASE_CONF_DIR="${HBASE_CONF_DIR:-$HBASE_HOME/conf}"
+
+if echo $HBASE_CONF_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+then
+  orig_hbaseconfdir=$HBASE_CONF_DIR
+  export HBASE_CONF_DIR=$(echo "$HBASE_CONF_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+fi
+
+if echo $HBASE_LOG_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+then
+  orig_hbaselogdir=$HBASE_LOG_DIR
+  export HBASE_LOG_DIR=$(echo "$HBASE_LOG_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+fi
+
+if echo $HBASE_PID_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+then
+  orig_hbasepiddir=$HBASE_PID_DIR
+  export HBASE_PID_DIR=$(echo "$HBASE_PID_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+fi
+
 # List of hbase regions servers.
 HBASE_REGIONSERVERS="${HBASE_REGIONSERVERS:-$HBASE_CONF_DIR/regionservers}"
 # List of hbase secondary masters.
diff -pruN hbase-0.98.17-hadoop2-alternate-ssh/bin/hbase-daemons.sh hbase-0.98.17-hadoop2/bin/hbase-daemons.sh
--- hbase-0.98.17-hadoop2-alternate-ssh/bin/hbase-daemons.sh	2016-01-07 16:17:02.000000000 -0800
+++ hbase-0.98.17-hadoop2/bin/hbase-daemons.sh	2016-08-04 13:19:57.610495000 -0700
@@ -35,6 +35,21 @@ bin=`cd "$bin">/dev/null; pwd`
 
 . $bin/hbase-config.sh
 
+if [ "${HBASE_CONF_DIR}X" != "X" ] && [ "${orig_hbaseconfdir}X" != "X" ]
+then
+  export HBASE_CONF_DIR=$orig_hbaseconfdir
+fi
+
+if [ "${HBASE_LOG_DIR}X" != "X" ] && [ "${orig_hbaselogdir}X" != "X" ]
+then
+  export HBASE_LOG_DIR=$orig_hbaselogdir
+fi
+
+if [ "${HBASE_PID_DIR}X" != "X" ] && [ "${orig_hbasepiddir}X" != "X" ]
+then
+  export HBASE_PID_DIR=$orig_hbasepiddir
+fi
+
 remote_cmd="cd ${HBASE_HOME}; $bin/hbase-daemon.sh --config ${HBASE_CONF_DIR} $@"
 args="--hosts ${HBASE_REGIONSERVERS} --config ${HBASE_CONF_DIR} $remote_cmd"
 
diff -pruN hbase-0.98.17-hadoop2-alternate-ssh/bin/start-hbase.sh hbase-0.98.17-hadoop2/bin/start-hbase.sh
--- hbase-0.98.17-hadoop2-alternate-ssh/bin/start-hbase.sh	2016-01-15 19:27:11.000000000 -0800
+++ hbase-0.98.17-hadoop2/bin/start-hbase.sh	2016-08-04 13:19:57.613492000 -0700
@@ -47,6 +47,20 @@ fi
 # HBASE-6504 - only take the first line of the output in case verbose gc is on
 distMode=`$bin/hbase --config "$HBASE_CONF_DIR" org.apache.hadoop.hbase.util.HBaseConfTool hbase.cluster.distributed | head -n 1`
 
+if [ "${HBASE_CONF_DIR}X" != "X" ] && [ "${orig_hbaseconfdir}X" != "X" ]
+then
+  export HBASE_CONF_DIR=$orig_hbaseconfdir
+fi
+
+if [ "${HBASE_LOG_DIR}X" != "X" ] && [ "${orig_hbaselogdir}X" != "X" ]
+then
+  export HBASE_LOG_DIR=$orig_hbaselogdir
+fi
+
+if [ "${HBASE_PID_DIR}X" != "X" ] && [ "${orig_hbasepiddir}X" != "X" ]
+then
+  export HBASE_PID_DIR=$orig_hbasepiddir
+fi
 
 if [ "$distMode" == 'false' ] 
 then
diff -pruN hbase-0.98.17-hadoop2-alternate-ssh/bin/stop-hbase. hbase-0.98.17-hadoop2/bin/stop-hbase.
--- hbase-0.98.17-hadoop2-alternate-ssh/bin/stop-hbase.	1969-12-31 16:00:00.000000000 -0800
+++ hbase-0.98.17-hadoop2/bin/stop-hbase.	2016-08-04 13:19:57.614492000 -0700
@@ -0,0 +1 @@
+s
\ No newline at end of file
diff -pruN hbase-0.98.17-hadoop2-alternate-ssh/bin/stop-hbase.sh hbase-0.98.17-hadoop2/bin/stop-hbase.sh
--- hbase-0.98.17-hadoop2-alternate-ssh/bin/stop-hbase.sh	2016-01-15 19:27:11.000000000 -0800
+++ hbase-0.98.17-hadoop2/bin/stop-hbase.sh	2016-08-04 13:19:57.617489000 -0700
@@ -59,6 +59,22 @@ rm -f $pid
 # distributed == false means that the HMaster will kill ZK when it exits
 # HBASE-6504 - only take the first line of the output in case verbose gc is on
 distMode=`$bin/hbase --config "$HBASE_CONF_DIR" org.apache.hadoop.hbase.util.HBaseConfTool hbase.cluster.distributed | head -n 1`
+
+if [ "${HBASE_CONF_DIR}X" != "X" ] && [ "${orig_hbaseconfdir}X" != "X" ]
+then
+  export HBASE_CONF_DIR=$orig_hbaseconfdir
+fi
+
+if [ "${HBASE_LOG_DIR}X" != "X" ] && [ "${orig_hbaselogdir}X" != "X" ]
+then
+  export HBASE_LOG_DIR=$orig_hbaselogdir
+fi
+
+if [ "${HBASE_PID_DIR}X" != "X" ] && [ "${orig_hbasepiddir}X" != "X" ]
+then
+  export HBASE_PID_DIR=$orig_hbasepiddir
+fi
+
 if [ "$distMode" == 'true' ] 
 then
   "$bin"/hbase-daemons.sh --config "${HBASE_CONF_DIR}" stop zookeeper
