diff -pruN hbase-0.98.8-hadoop2-orig/bin/hbase-cleanup.sh hbase-0.98.8-hadoop2-alternate-ssh/bin/hbase-cleanup.sh
--- hbase-0.98.8-hadoop2-orig/bin/hbase-cleanup.sh	2014-11-13 14:19:09.000000000 -0800
+++ hbase-0.98.8-hadoop2-alternate-ssh/bin/hbase-cleanup.sh	2016-08-04 11:53:55.961483000 -0700
@@ -28,6 +28,8 @@
 #   HBASE_CONF_DIR  Alternate hbase conf dir. Default is ${HBASE_HOME}/conf.
 #   HBASE_SLAVE_SLEEP Seconds to sleep between spawning remote commands.
 #   HBASE_SLAVE_TIMEOUT Seconds to wait for timing out a remote command.
+#   HBASE_SSH_CMD Specify an alternate remote shell command.
+#     Defaults to ssh if not specified.
 #   HBASE_SSH_OPTS Options passed to ssh when running remote commands.
 #
 
diff -pruN hbase-0.98.8-hadoop2-orig/bin/master-backup.sh hbase-0.98.8-hadoop2-alternate-ssh/bin/master-backup.sh
--- hbase-0.98.8-hadoop2-orig/bin/master-backup.sh	2014-11-13 14:19:09.000000000 -0800
+++ hbase-0.98.8-hadoop2-alternate-ssh/bin/master-backup.sh	2016-08-04 11:53:55.979467000 -0700
@@ -29,6 +29,8 @@
 #   HADOOP_CONF_DIR  Alternate conf dir. Default is ${HADOOP_HOME}/conf.
 #   HBASE_CONF_DIR  Alternate hbase conf dir. Default is ${HBASE_HOME}/conf.
 #   HBASE_SLAVE_SLEEP Seconds to sleep between spawning remote commands.
+#   HBASE_SSH_CMD Specify an alternate remote shell command.
+#     Defaults to ssh if not specified.
 #   HBASE_SSH_OPTS Options passed to ssh when running remote commands.
 #
 # Modelled after $HADOOP_HOME/bin/slaves.sh.
@@ -62,10 +64,11 @@ fi
 
 args=${@// /\\ }
 args=${args/master-backup/master}
+RSH_CMD=${HBASE_SSH_CMD:-ssh}
 
 if [ -f $HOSTLIST ]; then
   for hmaster in `cat "$HOSTLIST"`; do
-   ssh $HBASE_SSH_OPTS $hmaster $"$args --backup" \
+   $RSH_CMD $HBASE_SSH_OPTS $hmaster $"$args --backup" \
      2>&1 | sed "s/^/$hmaster: /" &
    if [ "$HBASE_SLAVE_SLEEP" != "" ]; then
      sleep $HBASE_SLAVE_SLEEP
diff -pruN hbase-0.98.8-hadoop2-orig/bin/regionservers.sh hbase-0.98.8-hadoop2-alternate-ssh/bin/regionservers.sh
--- hbase-0.98.8-hadoop2-orig/bin/regionservers.sh	2014-11-14 13:58:32.000000000 -0800
+++ hbase-0.98.8-hadoop2-alternate-ssh/bin/regionservers.sh	2016-08-04 11:53:55.997445000 -0700
@@ -29,6 +29,8 @@
 #   HADOOP_CONF_DIR  Alternate conf dir. Default is ${HADOOP_HOME}/conf.
 #   HBASE_CONF_DIR  Alternate hbase conf dir. Default is ${HBASE_HOME}/conf.
 #   HBASE_SLAVE_SLEEP Seconds to sleep between spawning remote commands.
+#   HBASE_SSH_CMD Specify an alternate remote shell command.
+#     Defaults to ssh if not specified.
 #   HBASE_SSH_OPTS Options passed to ssh when running remote commands.
 #
 # Modelled after $HADOOP_HOME/bin/slaves.sh.
@@ -50,6 +52,7 @@ bin=`cd "$bin">/dev/null; pwd`
 # then it takes precedence over the definition in 
 # hbase-env.sh. Save it here.
 HOSTLIST=$HBASE_REGIONSERVERS
+RSH_CMD=${HBASE_SSH_CMD:-ssh}
 
 if [ "$HOSTLIST" = "" ]; then
   if [ "$HBASE_REGIONSERVERS" = "" ]; then
@@ -61,10 +64,10 @@ fi
 
 for regionserver in `cat "$HOSTLIST"`; do
   if ${HBASE_SLAVE_PARALLEL:-true}; then 
-    ssh $HBASE_SSH_OPTS $regionserver $"${@// /\\ }" \
+    $RSH_CMD $HBASE_SSH_OPTS $regionserver $"${@// /\\ }" \
       2>&1 | sed "s/^/$regionserver: /" &
   else # run each command serially 
-    ssh $HBASE_SSH_OPTS $regionserver $"${@// /\\ }" \
+    $RSH_CMD $HBASE_SSH_OPTS $regionserver $"${@// /\\ }" \
       2>&1 | sed "s/^/$regionserver: /"
   fi
   if [ "$HBASE_SLAVE_SLEEP" != "" ]; then
diff -pruN hbase-0.98.8-hadoop2-orig/bin/rolling-restart.sh hbase-0.98.8-hadoop2-alternate-ssh/bin/rolling-restart.sh
--- hbase-0.98.8-hadoop2-orig/bin/rolling-restart.sh	2014-11-13 14:19:09.000000000 -0800
+++ hbase-0.98.8-hadoop2-alternate-ssh/bin/rolling-restart.sh	2016-08-04 11:53:56.000444000 -0700
@@ -30,6 +30,8 @@
 #   HBASE_CONF_DIR  Alternate hbase conf dir. Default is ${HBASE_HOME}/conf.
 #   HBASE_SLAVE_SLEEP Seconds to sleep between spawning remote commands.
 #   HBASE_SLAVE_TIMEOUT Seconds to wait for timing out a remote command. 
+#   HBASE_SSH_CMD Specify an alternate remote shell command.
+#     Defaults to ssh if not specified.
 #   HBASE_SSH_OPTS Options passed to ssh when running remote commands.
 #
 # Modelled after $HADOOP_HOME/bin/slaves.sh.
diff -pruN hbase-0.98.8-hadoop2-orig/bin/zookeepers.sh hbase-0.98.8-hadoop2-alternate-ssh/bin/zookeepers.sh
--- hbase-0.98.8-hadoop2-orig/bin/zookeepers.sh	2014-11-13 14:19:09.000000000 -0800
+++ hbase-0.98.8-hadoop2-alternate-ssh/bin/zookeepers.sh	2016-08-04 11:53:56.015427000 -0700
@@ -26,6 +26,8 @@
 #
 #   HBASE_CONF_DIR  Alternate hbase conf dir. Default is ${HBASE_HOME}/conf.
 #   HBASE_SLAVE_SLEEP Seconds to sleep between spawning remote commands.
+#   HADOOP_SSH_CMD Specify an alternate remote shell command.
+#     Defaults to ssh if not specified.
 #   HBASE_SSH_OPTS Options passed to ssh when running remote commands.
 #
 # Modelled after $HADOOP_HOME/bin/slaves.sh.
@@ -47,11 +49,13 @@ if [ "$HBASE_MANAGES_ZK" = "" ]; then
   HBASE_MANAGES_ZK=true
 fi
 
+RSH_CMD=${HBASE_SSH_CMD:-ssh}
+
 if [ "$HBASE_MANAGES_ZK" = "true" ]; then
   hosts=`"$bin"/hbase org.apache.hadoop.hbase.zookeeper.ZKServerTool | grep '^ZK host:' | sed 's,^ZK host:,,'`
   cmd=$"${@// /\\ }"
   for zookeeper in $hosts; do
-   ssh $HBASE_SSH_OPTS $zookeeper $cmd 2>&1 | sed "s/^/$zookeeper: /" &
+   $RSH_CMD $HBASE_SSH_OPTS $zookeeper $cmd 2>&1 | sed "s/^/$zookeeper: /" &
    if [ "$HBASE_SLAVE_SLEEP" != "" ]; then
      sleep $HBASE_SLAVE_SLEEP
    fi
