diff -pruN spark-1.2.1-bin-hadoop2.4-alternate-ssh/sbin/slaves.sh spark-1.2.1-bin-hadoop2.4/sbin/slaves.sh
--- spark-1.2.1-bin-hadoop2.4-alternate-ssh/sbin/slaves.sh	2015-09-24 11:00:18.557697000 -0700
+++ spark-1.2.1-bin-hadoop2.4/sbin/slaves.sh	2015-09-24 11:04:18.125889000 -0700
@@ -50,23 +50,61 @@ if [ -f "$SPARK_SLAVES" ]; then
   HOSTLIST=`cat "$SPARK_SLAVES"`
 fi
 
+myhostname=`hostname`
+
 # Check if --config is passed as an argument. It is an optional parameter.
 # Exit if the argument is not a directory.
 if [ "$1" == "--config" ]
 then
   shift
   conf_dir="$1"
+  if echo $conf_dir | grep -q MAGPIEHOSTNAMESUBSTITUTION
+  then
+      orig_conf_dir="$1"
+      conf_dir=$(echo "$conf_dir" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+  fi
   if [ ! -d "$conf_dir" ]
   then
     echo "ERROR : $conf_dir is not a directory"
     echo $usage
     exit 1
   else
+    if [ "${orig_conf_dir}X" != "X" ]
+    then
+	orig_sparkconfdir=$orig_conf_dir
+    fi
     export SPARK_CONF_DIR="$conf_dir"
   fi
   shift
 fi
 
+if [ "${SPARK_CONF_DIR}X" != "X" ]
+then
+    if echo $SPARK_CONF_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+        orig_sparkconfdir=$SPARK_CONF_DIR
+        export SPARK_CONF_DIR=$(echo "$SPARK_CONF_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+fi
+
+if [ "${SPARK_LOG_DIR}X" != "X" ]
+then
+    if echo $SPARK_LOG_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+        orig_sparklogdir=$SPARK_LOG_DIR
+        export SPARK_LOG_DIR=$(echo "$SPARK_LOG_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+fi
+
+if [ "${SPARK_PID_DIR}X" != "X" ]
+then
+    if echo $SPARK_PID_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+        orig_sparkpiddir=$SPARK_PID_DIR
+        export SPARK_PID_DIR=$(echo "$SPARK_PID_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+fi
+
 . "$SPARK_PREFIX/bin/load-spark-env.sh"
 
 if [ "$HOSTLIST" = "" ]; then
diff -pruN spark-1.2.1-bin-hadoop2.4-alternate-ssh/sbin/spark-daemon.sh spark-1.2.1-bin-hadoop2.4/sbin/spark-daemon.sh
--- spark-1.2.1-bin-hadoop2.4-alternate-ssh/sbin/spark-daemon.sh	2015-09-24 11:00:18.559697000 -0700
+++ spark-1.2.1-bin-hadoop2.4/sbin/spark-daemon.sh	2015-09-24 11:04:18.127890000 -0700
@@ -47,17 +47,28 @@ sbin="`cd "$sbin"; pwd`"
 # Check if --config is passed as an argument. It is an optional parameter.
 # Exit if the argument is not a directory.
 
+myhostname=`hostname`
+
 if [ "$1" == "--config" ]
 then
   shift
-  conf_dir="$1"
+  conf_dir=$1
+  if echo $conf_dir | grep -q MAGPIEHOSTNAMESUBSTITUTION
+  then
+      orig_conf_dir="$1"
+      conf_dir=$(echo "$conf_dir" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+  fi
   if [ ! -d "$conf_dir" ]
   then
     echo "ERROR : $conf_dir is not a directory"
     echo $usage
     exit 1
   else
-    export SPARK_CONF_DIR="$conf_dir"
+    if [ "${orig_conf_dir}X" != "X" ]
+    then
+        orig_sparkconfdir=$orig_conf_dir
+    fi
+    export SPARK_CONF_DIR=$conf_dir
   fi
   shift
 fi
@@ -69,6 +80,33 @@ shift
 instance=$1
 shift
 
+if [ "${SPARK_CONF_DIR}X" != "X" ]
+then
+    if echo $SPARK_CONF_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+        orig_sparkconfdir=$SPARK_CONF_DIR
+        export SPARK_CONF_DIR=$(echo "$SPARK_CONF_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+fi
+
+if [ "${SPARK_LOG_DIR}X" != "X" ]
+then
+    if echo $SPARK_LOG_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+        orig_sparklogdir=$SPARK_LOG_DIR
+        export SPARK_LOG_DIR=$(echo "$SPARK_LOG_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+fi
+
+if [ "${SPARK_PID_DIR}X" != "X" ]
+then
+    if echo $SPARK_PID_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+        orig_sparkpiddir=$SPARK_PID_DIR
+        export SPARK_PID_DIR=$(echo "$SPARK_PID_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+fi
+
 spark_rotate_log ()
 {
     log=$1;
diff -pruN spark-1.2.1-bin-hadoop2.4-alternate-ssh/sbin/spark-daemons.sh spark-1.2.1-bin-hadoop2.4/sbin/spark-daemons.sh
--- spark-1.2.1-bin-hadoop2.4-alternate-ssh/sbin/spark-daemons.sh	2015-09-24 11:00:07.995270000 -0700
+++ spark-1.2.1-bin-hadoop2.4/sbin/spark-daemons.sh	2015-09-24 11:04:18.130886000 -0700
@@ -30,6 +30,8 @@ fi
 sbin=`dirname "$0"`
 sbin=`cd "$sbin"; pwd`
 
+myhostname=`hostname`
+
 # Check if --config is passed as an argument. It is an optional parameter.
 # Exit if the argument is not a directory.
 
@@ -37,12 +39,21 @@ if [ "$1" == "--config" ]
 then
   shift
   conf_dir=$1
+  if echo $conf_dir | grep -q MAGPIEHOSTNAMESUBSTITUTION
+  then
+      orig_conf_dir="$1"
+      conf_dir=$(echo "$conf_dir" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+  fi
   if [ ! -d "$conf_dir" ]
   then
     echo "ERROR : $conf_dir is not a directory"
     echo $usage
     exit 1
   else
+    if [ "${orig_conf_dir}X" != "X" ]
+    then
+        orig_sparkconfdir=$orig_conf_dir
+    fi
     export SPARK_CONF_DIR=$conf_dir
   fi
   shift
@@ -50,4 +61,18 @@ fi
 
 . "$sbin/spark-config.sh"
 
+if [ "${SPARK_CONF_DIR}X" != "X" ]
+then
+    if echo $SPARK_CONF_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+        orig_sparkconfdir=$SPARK_CONF_DIR
+        export SPARK_CONF_DIR=$(echo "$SPARK_CONF_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+fi
+
+if [ "${SPARK_CONF_DIR}X" != "X" ] && [ "${orig_sparkconfdir}X" != "X" ]
+then
+    export SPARK_CONF_DIR=$orig_sparkconfdir
+fi
+
 exec "$sbin/slaves.sh" --config $SPARK_CONF_DIR cd "$SPARK_HOME" \; "$sbin/spark-daemon.sh" --config $SPARK_CONF_DIR "$@"
diff -pruN spark-1.2.1-bin-hadoop2.4-alternate-ssh/sbin/start-master.sh spark-1.2.1-bin-hadoop2.4/sbin/start-master.sh
--- spark-1.2.1-bin-hadoop2.4-alternate-ssh/sbin/start-master.sh	2015-02-02 19:45:55.000000000 -0800
+++ spark-1.2.1-bin-hadoop2.4/sbin/start-master.sh	2015-09-24 11:04:18.133882000 -0700
@@ -24,6 +24,8 @@ sbin="`cd "$sbin"; pwd`"
 
 START_TACHYON=false
 
+myhostname=`hostname`
+	
 while (( "$#" )); do
 case $1 in
     --with-tachyon)
@@ -39,6 +41,33 @@ done
 
 . "$sbin/spark-config.sh"
 
+if [ "${SPARK_CONF_DIR}X" != "X" ]
+then
+    if echo $SPARK_CONF_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+	orig_sparkconfdir=$SPARK_CONF_DIR
+	export SPARK_CONF_DIR=$(echo "$SPARK_CONF_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+fi
+
+if [ "${SPARK_LOG_DIR}X" != "X" ]
+then
+    if echo $SPARK_LOG_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+	orig_sparklogdir=$SPARK_LOG_DIR
+	export SPARK_LOG_DIR=$(echo "$SPARK_LOG_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+fi
+
+if [ "${SPARK_PID_DIR}X" != "X" ]
+then
+    if echo $SPARK_PID_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+	orig_sparkpiddir=$SPARK_PID_DIR
+	export SPARK_PID_DIR=$(echo "$SPARK_PID_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+fi
+
 . "$SPARK_PREFIX/bin/load-spark-env.sh"
 
 if [ "$SPARK_MASTER_PORT" = "" ]; then
@@ -53,6 +82,21 @@ if [ "$SPARK_MASTER_WEBUI_PORT" = "" ]; 
   SPARK_MASTER_WEBUI_PORT=8080
 fi
 
+if [ "${SPARK_CONF_DIR}X" != "X" ] && [ "${orig_sparkconfdir}X" != "X" ]
+then
+    export SPARK_CONF_DIR=$orig_sparkconfdir
+fi
+
+if [ "${SPARK_LOG_DIR}X" != "X" ] && [ "${orig_sparklogdir}X" != "X" ]
+then
+    export SPARK_LOG_DIR=$orig_sparklogdir
+fi
+
+if [ "${SPARK_PID_DIR}X" != "X" ] && [ "${orig_sparkpiddir}X" != "X" ]
+then
+    export SPARK_PID_DIR=$orig_sparkpiddir
+fi
+
 "$sbin"/spark-daemon.sh start org.apache.spark.deploy.master.Master 1 --ip $SPARK_MASTER_IP --port $SPARK_MASTER_PORT --webui-port $SPARK_MASTER_WEBUI_PORT
 
 if [ "$START_TACHYON" == "true" ]; then
diff -pruN spark-1.2.1-bin-hadoop2.4-alternate-ssh/sbin/start-slave.sh spark-1.2.1-bin-hadoop2.4/sbin/start-slave.sh
--- spark-1.2.1-bin-hadoop2.4-alternate-ssh/sbin/start-slave.sh	2015-09-24 11:00:07.997269000 -0700
+++ spark-1.2.1-bin-hadoop2.4/sbin/start-slave.sh	2015-09-24 11:04:18.136878000 -0700
@@ -26,16 +26,27 @@ sbin="`cd "$sbin"; pwd`"
 # Check if --config is passed as an argument. It is an optional parameter.
 # Exit if the argument is not a directory.
 
+myhostname=`hostname`
+
 if [ "$1" == "--config" ]
 then
   shift
   conf_dir=$1
+  if echo $conf_dir | grep -q MAGPIEHOSTNAMESUBSTITUTION
+  then
+      orig_conf_dir="$1"
+      conf_dir=$(echo "$conf_dir" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+  fi
   if [ ! -d "$conf_dir" ]
   then
     echo "ERROR : $conf_dir is not a directory"
     echo $usage
     exit 1
   else
+    if [ "${orig_conf_dir}X" != "X" ]
+    then
+        orig_sparkconfdir=$orig_conf_dir
+    fi
     export SPARK_CONF_DIR=$conf_dir
   fi
   shift
diff -pruN spark-1.2.1-bin-hadoop2.4-alternate-ssh/sbin/start-slaves.sh spark-1.2.1-bin-hadoop2.4/sbin/start-slaves.sh
--- spark-1.2.1-bin-hadoop2.4-alternate-ssh/sbin/start-slaves.sh	2015-09-24 11:00:08.000266000 -0700
+++ spark-1.2.1-bin-hadoop2.4/sbin/start-slaves.sh	2015-09-24 11:04:18.138881000 -0700
@@ -23,6 +23,8 @@ sbin="`cd "$sbin"; pwd`"
 
 START_TACHYON=false
 
+myhostname=`hostname`
+
 while (( "$#" )); do
 case $1 in
     --with-tachyon)
@@ -38,6 +40,33 @@ done
 
 . "$sbin/spark-config.sh"
 
+if [ "${SPARK_CONF_DIR}X" != "X" ]
+then
+    if echo $SPARK_CONF_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+	orig_sparkconfdir=$SPARK_CONF_DIR
+	export SPARK_CONF_DIR=$(echo "$SPARK_CONF_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+fi
+
+if [ "${SPARK_LOG_DIR}X" != "X" ]
+then
+    if echo $SPARK_LOG_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+	orig_sparklogdir=$SPARK_LOG_DIR
+	export SPARK_LOG_DIR=$(echo "$SPARK_LOG_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+fi
+
+if [ "${SPARK_PID_DIR}X" != "X" ]
+then
+    if echo $SPARK_PID_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+	orig_sparkpiddir=$SPARK_PID_DIR
+	export SPARK_PID_DIR=$(echo "$SPARK_PID_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+fi
+
 . "$SPARK_PREFIX/bin/load-spark-env.sh"
 
 # Find the port number for the master
@@ -49,6 +78,21 @@ if [ "$SPARK_MASTER_IP" = "" ]; then
   SPARK_MASTER_IP="`hostname`"
 fi
 
+if [ "${SPARK_CONF_DIR}X" != "X" ] && [ "${orig_sparkconfdir}X" != "X" ]
+then
+    export SPARK_CONF_DIR=$orig_sparkconfdir
+fi
+
+if [ "${SPARK_LOG_DIR}X" != "X" ] && [ "${orig_sparklogdir}X" != "X" ]
+then
+    export SPARK_LOG_DIR=$orig_sparklogdir
+fi
+
+if [ "${SPARK_PID_DIR}X" != "X" ] && [ "${orig_sparkpiddir}X" != "X" ]
+then
+    export SPARK_PID_DIR=$orig_sparkpiddir
+fi
+
 if [ "$START_TACHYON" == "true" ]; then
   "$sbin/slaves.sh" cd "$SPARK_HOME" \; "$sbin"/../tachyon/bin/tachyon bootstrap-conf "$SPARK_MASTER_IP"
 
diff -pruN spark-1.2.1-bin-hadoop2.4-alternate-ssh/sbin/stop-master.sh spark-1.2.1-bin-hadoop2.4/sbin/stop-master.sh
--- spark-1.2.1-bin-hadoop2.4-alternate-ssh/sbin/stop-master.sh	2015-02-02 19:45:55.000000000 -0800
+++ spark-1.2.1-bin-hadoop2.4/sbin/stop-master.sh	2015-09-24 11:04:18.141875000 -0700
@@ -22,8 +22,24 @@
 sbin=`dirname "$0"`
 sbin=`cd "$sbin"; pwd`
 
+myhostname=`hostname`
+
 . "$sbin/spark-config.sh"
 
+if [ "${SPARK_CONF_DIR}X" != "X" ]
+then
+    if echo $SPARK_CONF_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+        orig_sparkconfdir=$SPARK_CONF_DIR
+        export SPARK_CONF_DIR=$(echo "$SPARK_CONF_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+fi
+
+if [ "${SPARK_CONF_DIR}X" != "X" ] && [ "${orig_sparkconfdir}X" != "X" ]
+then
+    export SPARK_CONF_DIR=$orig_sparkconfdir
+fi
+
 "$sbin"/spark-daemon.sh stop org.apache.spark.deploy.master.Master 1
 
 if [ -e "$sbin"/../tachyon/bin/tachyon ]; then
diff -pruN spark-1.2.1-bin-hadoop2.4-alternate-ssh/sbin/stop-slaves.sh spark-1.2.1-bin-hadoop2.4/sbin/stop-slaves.sh
--- spark-1.2.1-bin-hadoop2.4-alternate-ssh/sbin/stop-slaves.sh	2015-09-24 11:00:08.003262000 -0700
+++ spark-1.2.1-bin-hadoop2.4/sbin/stop-slaves.sh	2015-09-24 11:04:18.144871000 -0700
@@ -20,10 +20,54 @@
 sbin=`dirname "$0"`
 sbin=`cd "$sbin"; pwd`
 
+myhostname=`hostname`
+
 . "$sbin/spark-config.sh"
 
+if [ "${SPARK_CONF_DIR}X" != "X" ]
+then
+    if echo $SPARK_CONF_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+        orig_sparkconfdir=$SPARK_CONF_DIR
+        export SPARK_CONF_DIR=$(echo "$SPARK_CONF_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+fi
+
+if [ "${SPARK_LOG_DIR}X" != "X" ]
+then
+    if echo $SPARK_LOG_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+        orig_sparklogdir=$SPARK_LOG_DIR
+        export SPARK_LOG_DIR=$(echo "$SPARK_LOG_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+fi
+
+if [ "${SPARK_PID_DIR}X" != "X" ]
+then
+    if echo $SPARK_PID_DIR | grep -q MAGPIEHOSTNAMESUBSTITUTION
+    then
+        orig_sparkpiddir=$SPARK_PID_DIR
+        export SPARK_PID_DIR=$(echo "$SPARK_PID_DIR" | sed "s/MAGPIEHOSTNAMESUBSTITUTION/$myhostname/g")
+    fi
+fi
+
 . "$SPARK_PREFIX/bin/load-spark-env.sh"
 
+if [ "${SPARK_CONF_DIR}X" != "X" ] && [ "${orig_sparkconfdir}X" != "X" ]
+then
+    export SPARK_CONF_DIR=$orig_sparkconfdir
+fi
+
+if [ "${SPARK_LOG_DIR}X" != "X" ] && [ "${orig_sparklogdir}X" != "X" ]
+then
+    export SPARK_LOG_DIR=$orig_sparklogdir
+fi
+
+if [ "${SPARK_PID_DIR}X" != "X" ] && [ "${orig_sparkpiddir}X" != "X" ]
+then
+    export SPARK_PID_DIR=$orig_sparkpiddir
+fi
+
 # do before the below calls as they exec
 if [ -e "$sbin"/../tachyon/bin/tachyon ]; then
   "$sbin/slaves.sh" cd "$SPARK_HOME" \; "$sbin"/../tachyon/bin/tachyon killAll tachyon.worker.Worker
