diff -pruN spark-1.2.0-bin-hadoop2.4-orig/sbin/slaves.sh spark-1.2.0-bin-hadoop2.4/sbin/slaves.sh
--- spark-1.2.0-bin-hadoop2.4-orig/sbin/slaves.sh	2014-12-10 03:00:25.000000000 -0800
+++ spark-1.2.0-bin-hadoop2.4/sbin/slaves.sh	2015-01-20 11:17:42.478717000 -0800
@@ -25,6 +25,8 @@
 #     Default is ${SPARK_CONF_DIR}/slaves.
 #   SPARK_CONF_DIR  Alternate conf dir. Default is ${SPARK_HOME}/conf.
 #   SPARK_SLAVE_SLEEP Seconds to sleep between spawning remote commands.
+#   SPARK_SSH_CMD Specify an alternate remote shell command.
+#     Defaults to ssh if not specified.
 #   SPARK_SSH_OPTS Options passed to ssh when running remote commands.
 ##
 
@@ -79,19 +81,19 @@ if [ "$HOSTLIST" = "" ]; then
   fi
 fi
 
-
+RSH_CMD=${SPARK_SSH_CMD:-ssh}
 
 # By default disable strict host key checking
-if [ "$SPARK_SSH_OPTS" = "" ]; then
+if [ "$RSH_CMD" == "ssh" ] && [ "$SPARK_SSH_OPTS" = "" ]; then
   SPARK_SSH_OPTS="-o StrictHostKeyChecking=no"
 fi
 
 for slave in `echo "$HOSTLIST"|sed  "s/#.*$//;/^$/d"`; do
   if [ -n "${SPARK_SSH_FOREGROUND}" ]; then
-    ssh $SPARK_SSH_OPTS "$slave" $"${@// /\\ }" \
+    $RSH_CMD $SPARK_SSH_OPTS "$slave" $"${@// /\\ }" \
       2>&1 | sed "s/^/$slave: /"
   else
-    ssh $SPARK_SSH_OPTS "$slave" $"${@// /\\ }" \
+    $RSH_CMD $SPARK_SSH_OPTS "$slave" $"${@// /\\ }" \
       2>&1 | sed "s/^/$slave: /" &
   fi
   if [ "$SPARK_SLAVE_SLEEP" != "" ]; then
diff -pruN spark-1.2.0-bin-hadoop2.4-orig/sbin/spark-daemon.sh spark-1.2.0-bin-hadoop2.4/sbin/spark-daemon.sh
--- spark-1.2.0-bin-hadoop2.4-orig/sbin/spark-daemon.sh	2014-12-10 03:00:25.000000000 -0800
+++ spark-1.2.0-bin-hadoop2.4/sbin/spark-daemon.sh	2015-01-20 11:18:19.185971000 -0800
@@ -137,7 +137,8 @@ case $option in
 
     if [ "$SPARK_MASTER" != "" ]; then
       echo rsync from "$SPARK_MASTER"
-      rsync -a -e ssh --delete --exclude=.svn --exclude='logs/*' --exclude='contrib/hod/logs/*' $SPARK_MASTER/ "$SPARK_HOME"
+      RSH_CMD=${SPARK_SSH_CMD:-ssh}
+      rsync -a -e $RSH_CMD --delete --exclude=.svn --exclude='logs/*' --exclude='contrib/hod/logs/*' $SPARK_MASTER/ "$SPARK_HOME"
     fi
 
     spark_rotate_log "$log"
