diff -pruN spark-0.9.1-orig/sbin/slaves.sh spark-0.9.1/sbin/slaves.sh
--- spark-0.9.1-orig/sbin/slaves.sh	2014-03-26 22:28:16.000000000 -0700
+++ spark-0.9.1/sbin/slaves.sh	2014-04-14 15:45:11.290667000 -0700
@@ -25,6 +25,8 @@
 #     Default is ${SPARK_CONF_DIR}/slaves.
 #   SPARK_CONF_DIR  Alternate conf dir. Default is ${SPARK_HOME}/conf.
 #   SPARK_SLAVE_SLEEP Seconds to sleep between spawning remote commands.
+#   SPARK_SSH_CMD Specify an alternate remote shell command.
+#     Defaults to ssh if not specified.
 #   SPARK_SSH_OPTS Options passed to ssh when running remote commands.
 ##
 
@@ -75,13 +77,15 @@ if [ "$HOSTLIST" = "" ]; then
   fi
 fi
 
+RSH_CMD=${SPARK_SSH_CMD:-ssh}
+
 # By default disable strict host key checking
-if [ "$SPARK_SSH_OPTS" = "" ]; then
+if [ "$RSH_CMD" == "ssh" ] && [ "$SPARK_SSH_OPTS" = "" ]; then
   SPARK_SSH_OPTS="-o StrictHostKeyChecking=no"
 fi
 
 for slave in `cat "$HOSTLIST"|sed  "s/#.*$//;/^$/d"`; do
- ssh $SPARK_SSH_OPTS $slave $"${@// /\\ }" \
+ $RSH_CMD $SPARK_SSH_OPTS $slave $"${@// /\\ }" \
    2>&1 | sed "s/^/$slave: /" &
  if [ "$SPARK_SLAVE_SLEEP" != "" ]; then
    sleep $SPARK_SLAVE_SLEEP
diff -pruN spark-0.9.1-orig/sbin/spark-daemon.sh spark-0.9.1/sbin/spark-daemon.sh
--- spark-0.9.1-orig/sbin/spark-daemon.sh	2014-03-26 22:28:16.000000000 -0700
+++ spark-0.9.1/sbin/spark-daemon.sh	2014-04-14 15:33:25.945718000 -0700
@@ -141,7 +141,8 @@ case $startStop in
 
     if [ "$SPARK_MASTER" != "" ]; then
       echo rsync from $SPARK_MASTER
-      rsync -a -e ssh --delete --exclude=.svn --exclude='logs/*' --exclude='contrib/hod/logs/*' $SPARK_MASTER/ "$SPARK_HOME"
+      RSH_CMD=${SPARK_SSH_CMD:-ssh}
+      rsync -a -e $RSH_CMD --delete --exclude=.svn --exclude='logs/*' --exclude='contrib/hod/logs/*' $SPARK_MASTER/ "$SPARK_HOME"
     fi
 
     spark_rotate_log "$log"
