diff -pruN apache-phoenix-4.8.1-HBase-1.0-bin-orig/bin/performance.py apache-phoenix-4.8.1-HBase-1.0-bin/bin/performance.py
--- apache-phoenix-4.8.1-HBase-1.0-bin-orig/bin/performance.py	2016-09-22 18:00:49.000000000 -0700
+++ apache-phoenix-4.8.1-HBase-1.0-bin/bin/performance.py	2016-10-01 08:42:01.229359000 -0700
@@ -128,7 +128,7 @@ queryex("4 - Truncate + Group By", "SELE
 queryex("5 - Filter + Count", "SELECT COUNT(1) FROM %s WHERE CORE<10;" % (table))
 
 print "\nGenerating and upserting data..."
-exitcode = subprocess.call('java -jar %s %s %s' % (phoenix_utils.testjar, data, rowcount),
+exitcode = subprocess.call('%s -jar %s %s %s' % (java_cmd, phoenix_utils.testjar, data, rowcount),
                            shell=True)
 if exitcode != 0:
     sys.exit(exitcode)
