diff -pruN hadoop-2.7.6-orig/sbin/hadoop-daemon.sh hadoop-2.7.6-alternate-ssh/sbin/hadoop-daemon.sh
--- hadoop-2.7.6-orig/sbin/hadoop-daemon.sh	2018-04-17 18:39:20.000000000 -0700
+++ hadoop-2.7.6-alternate-ssh/sbin/hadoop-daemon.sh	2019-01-25 10:33:55.265928000 -0800
@@ -26,6 +26,7 @@
 #   HADOOP_PID_DIR   The pid files are stored. /tmp by default.
 #   HADOOP_IDENT_STRING   A string representing this instance of hadoop. $USER by default
 #   HADOOP_NICENESS The scheduling priority for daemons. Defaults to 0.
+#   HADOOP_SSH_CMD   Specify an alternate remote shell comand
 ##
 
 usage="Usage: hadoop-daemon.sh [--config <conf-dir>] [--hosts hostlistfile] [--script script] (start|stop) <hadoop-command> <args...>"
@@ -122,6 +123,7 @@ export HDFS_AUDIT_LOGGER=${HDFS_AUDIT_LO
 log=$HADOOP_LOG_DIR/hadoop-$HADOOP_IDENT_STRING-$command-$HOSTNAME.out
 pid=$HADOOP_PID_DIR/hadoop-$HADOOP_IDENT_STRING-$command.pid
 HADOOP_STOP_TIMEOUT=${HADOOP_STOP_TIMEOUT:-5}
+RSH_CMD=${HADOOP_SSH_CMD:-ssh}
 
 # Set default scheduling priority
 if [ "$HADOOP_NICENESS" = "" ]; then
@@ -143,7 +145,7 @@ case $startStop in
 
     if [ "$HADOOP_MASTER" != "" ]; then
       echo rsync from $HADOOP_MASTER
-      rsync -a -e ssh --delete --exclude=.svn --exclude='logs/*' --exclude='contrib/hod/logs/*' $HADOOP_MASTER/ "$HADOOP_PREFIX"
+      rsync -a -e $RSH_CMD --delete --exclude=.svn --exclude='logs/*' --exclude='contrib/hod/logs/*' $HADOOP_MASTER/ "$HADOOP_PREFIX"
     fi
 
     hadoop_rotate_log $log
diff -pruN hadoop-2.7.6-orig/sbin/slaves.sh hadoop-2.7.6-alternate-ssh/sbin/slaves.sh
--- hadoop-2.7.6-orig/sbin/slaves.sh	2018-04-17 18:39:20.000000000 -0700
+++ hadoop-2.7.6-alternate-ssh/sbin/slaves.sh	2019-01-25 10:33:55.288025000 -0800
@@ -24,6 +24,8 @@
 #     Default is ${HADOOP_CONF_DIR}/slaves.
 #   HADOOP_CONF_DIR  Alternate conf dir. Default is ${HADOOP_PREFIX}/conf.
 #   HADOOP_SLAVE_SLEEP Seconds to sleep between spawning remote commands.
+#   HADOOP_SSH_CMD Specify an alternate remote shell command.
+#     Defaults to ssh if not specified.
 #   HADOOP_SSH_OPTS Options passed to ssh when running remote commands.
 ##
 
@@ -55,9 +57,11 @@ else
   SLAVE_NAMES=$(cat "$SLAVE_FILE" | sed  's/#.*$//;/^$/d')
 fi
 
+RSH_CMD=${HADOOP_SSH_CMD:-ssh}
+
 # start the daemons
 for slave in $SLAVE_NAMES ; do
- ssh $HADOOP_SSH_OPTS $slave $"${@// /\\ }" \
+ $RSH_CMD $HADOOP_SSH_OPTS $slave $"${@// /\\ }" \
    2>&1 | sed "s/^/$slave: /" &
  if [ "$HADOOP_SLAVE_SLEEP" != "" ]; then
    sleep $HADOOP_SLAVE_SLEEP
diff -pruN hadoop-2.7.6-orig/sbin/yarn-daemon.sh hadoop-2.7.6-alternate-ssh/sbin/yarn-daemon.sh
--- hadoop-2.7.6-orig/sbin/yarn-daemon.sh	2018-04-17 18:39:20.000000000 -0700
+++ hadoop-2.7.6-alternate-ssh/sbin/yarn-daemon.sh	2019-01-25 10:33:55.303084000 -0800
@@ -26,6 +26,7 @@
 #   YARN_PID_DIR   The pid files are stored. /tmp by default.
 #   YARN_IDENT_STRING   A string representing this instance of hadoop. $USER by default
 #   YARN_NICENESS The scheduling priority for daemons. Defaults to 0.
+#   YARN_SSH_CMD   Specify an alternate remote shell comand
 ##
 
 usage="Usage: yarn-daemon.sh [--config <conf-dir>] [--hosts hostlistfile] (start|stop) <yarn-command> "
@@ -94,6 +95,7 @@ export YARN_ROOT_LOGGER=${YARN_ROOT_LOGG
 log=$YARN_LOG_DIR/yarn-$YARN_IDENT_STRING-$command-$HOSTNAME.out
 pid=$YARN_PID_DIR/yarn-$YARN_IDENT_STRING-$command.pid
 YARN_STOP_TIMEOUT=${YARN_STOP_TIMEOUT:-5}
+RSH_CMD=${YARN_SSH_CMD:-ssh}
 
 # Set default scheduling priority
 if [ "$YARN_NICENESS" = "" ]; then
@@ -115,7 +117,7 @@ case $startStop in
 
     if [ "$YARN_MASTER" != "" ]; then
       echo rsync from $YARN_MASTER
-      rsync -a -e ssh --delete --exclude=.svn --exclude='logs/*' --exclude='contrib/hod/logs/*' $YARN_MASTER/ "$HADOOP_YARN_HOME"
+      rsync -a -e $RSH_CMD --delete --exclude=.svn --exclude='logs/*' --exclude='contrib/hod/logs/*' $YARN_MASTER/ "$HADOOP_YARN_HOME"
     fi
 
     hadoop_rotate_log $log
