diff -pruN hadoop-1.2.1-orig/bin/hadoop-daemon.sh hadoop-1.2.1/bin/hadoop-daemon.sh
--- hadoop-1.2.1-orig/bin/hadoop-daemon.sh	2013-07-22 15:26:38.000000000 -0700
+++ hadoop-1.2.1/bin/hadoop-daemon.sh	2016-07-15 10:21:01.625158000 -0700
@@ -26,6 +26,7 @@
 #   HADOOP_PID_DIR   The pid files are stored. /tmp by default.
 #   HADOOP_IDENT_STRING   A string representing this instance of hadoop. $USER by default
 #   HADOOP_NICENESS The scheduling priority for daemons. Defaults to 0.
+#   HADOOP_SSH_CMD   Specify an alternate remote shell comand
 ##
 
 usage="Usage: hadoop-daemon.sh [--config <conf-dir>] [--hosts hostlistfile] (start|stop) <hadoop-command> <args...>"
@@ -107,6 +108,7 @@ export HADOOP_ROOT_LOGGER="INFO,DRFA"
 log=$HADOOP_LOG_DIR/hadoop-$HADOOP_IDENT_STRING-$command-$HOSTNAME.out
 pid=$HADOOP_PID_DIR/hadoop-$HADOOP_IDENT_STRING-$command.pid
 HADOOP_STOP_TIMEOUT=${HADOOP_STOP_TIMEOUT:-5}
+RSH_CMD=${HADOOP_SSH_CMD:-ssh}
 
 # Set default scheduling priority
 if [ "$HADOOP_NICENESS" = "" ]; then
@@ -128,7 +130,7 @@ case $startStop in
 
     if [ "$HADOOP_MASTER" != "" ]; then
       echo rsync from $HADOOP_MASTER
-      rsync -a -e ssh --delete --exclude=.svn --exclude='logs/*' --exclude='contrib/hod/logs/*' $HADOOP_MASTER/ "$HADOOP_HOME"
+      rsync -a -e $RSH_CMD --delete --exclude=.svn --exclude='logs/*' --exclude='contrib/hod/logs/*' $HADOOP_MASTER/ "$HADOOP_HOME"
     fi
 
     hadoop_rotate_log $log
diff -pruN hadoop-1.2.1-orig/bin/slaves.sh hadoop-1.2.1/bin/slaves.sh
--- hadoop-1.2.1-orig/bin/slaves.sh	2013-07-22 15:26:38.000000000 -0700
+++ hadoop-1.2.1/bin/slaves.sh	2016-07-15 10:21:01.627159000 -0700
@@ -24,6 +24,8 @@
 #     Default is ${HADOOP_CONF_DIR}/slaves.
 #   HADOOP_CONF_DIR  Alternate conf dir. Default is ${HADOOP_HOME}/conf.
 #   HADOOP_SLAVE_SLEEP Seconds to sleep between spawning remote commands.
+#   HADOOP_SSH_CMD Specify an alternate remote shell command.
+#     Defaults to ssh if not specified.
 #   HADOOP_SSH_OPTS Options passed to ssh when running remote commands.
 ##
 
@@ -58,8 +60,10 @@ if [ "$HOSTLIST" = "" ]; then
   fi
 fi
 
+RSH_CMD=${HADOOP_SSH_CMD:-ssh}
+
 for slave in `cat "$HOSTLIST"|sed  "s/#.*$//;/^$/d"`; do
- ssh $HADOOP_SSH_OPTS $slave $"${@// /\\ }" \
+ $RSH_CMD $HADOOP_SSH_OPTS $slave $"${@// /\\ }" \
    2>&1 | sed "s/^/$slave: /" &
  if [ "$HADOOP_SLAVE_SLEEP" != "" ]; then
    sleep $HADOOP_SLAVE_SLEEP
