diff -pruN apache-mahout-distribution-0.11.1-orig/examples/bin/classify-20newsgroups.sh apache-mahout-distribution-0.11.1/examples/bin/classify-20newsgroups.sh
--- apache-mahout-distribution-0.11.1-orig/examples/bin/classify-20newsgroups.sh	2015-11-06 11:14:37.000000000 -0800
+++ apache-mahout-distribution-0.11.1/examples/bin/classify-20newsgroups.sh	2015-11-19 13:47:17.293792000 -0800
@@ -36,7 +36,11 @@ START_PATH=`pwd`
 # Set commands for dfs
 source ${START_PATH}/set-dfs-commands.sh
 
-WORK_DIR=/tmp/mahout-work-${USER}
+if [[ -z "$MAHOUT_WORK_DIR" ]]; then
+  WORK_DIR=/tmp/mahout-work-${USER}
+else
+  WORK_DIR=$MAHOUT_WORK_DIR
+fi
 algorithm=( cnaivebayes-MapReduce naivebayes-MapReduce cnaivebayes-Spark naivebayes-Spark sgd clean)
 if [ -n "$1" ]; then
   choice=$1
@@ -105,7 +109,7 @@ if  ( [ "x$alg" == "xnaivebayes-MapReduc
     echo "Copying 20newsgroups data to HDFS"
     set +e
     $DFSRM ${WORK_DIR}/20news-all
-    $DFS -mkdir ${WORK_DIR}
+    $DFS -mkdir -p ${WORK_DIR}
     $DFS -mkdir ${WORK_DIR}/20news-all
     set -e
     if [ $HVERSION -eq "1" ] ; then
diff -pruN apache-mahout-distribution-0.11.1-orig/examples/bin/classify-wikipedia.sh apache-mahout-distribution-0.11.1/examples/bin/classify-wikipedia.sh
--- apache-mahout-distribution-0.11.1-orig/examples/bin/classify-wikipedia.sh	2015-11-06 11:14:37.000000000 -0800
+++ apache-mahout-distribution-0.11.1/examples/bin/classify-wikipedia.sh	2015-11-19 13:47:17.295790000 -0800
@@ -42,7 +42,11 @@ START_PATH=`pwd`
 # Set commands for dfs
 source ${START_PATH}/set-dfs-commands.sh
 
-WORK_DIR=/tmp/mahout-work-wiki
+if [[ -z "$MAHOUT_WORK_DIR" ]]; then
+  WORK_DIR=/tmp/mahout-work-wiki
+else
+  WORK_DIR=$MAHOUT_WORK_DIR
+fi
 algorithm=( CBayes BinaryCBayes clean)
 if [ -n "$1" ]; then
   choice=$1
@@ -110,7 +114,7 @@ if [ "x$alg" == "xCBayes" ] || [ "x$alg"
     echo "Copying wikipedia data to HDFS"
     set +e
     $DFSRM ${WORK_DIR}/wikixml
-    $DFS -mkdir ${WORK_DIR}
+    $DFS -mkdir -p ${WORK_DIR}
     set -e
     $DFS -put ${WORK_DIR}/wikixml ${WORK_DIR}/wikixml
   fi
diff -pruN apache-mahout-distribution-0.11.1-orig/examples/bin/cluster-reuters.sh apache-mahout-distribution-0.11.1/examples/bin/cluster-reuters.sh
--- apache-mahout-distribution-0.11.1-orig/examples/bin/cluster-reuters.sh	2015-11-06 11:14:37.000000000 -0800
+++ apache-mahout-distribution-0.11.1/examples/bin/cluster-reuters.sh	2015-11-19 13:47:17.298785000 -0800
@@ -43,6 +43,12 @@ if [ ! -e $MAHOUT ]; then
   exit 1
 fi
 
+if [[ -z "$MAHOUT_WORK_DIR" ]]; then
+  WORK_DIR=/tmp/mahout-work-${USER}
+else
+  WORK_DIR=$MAHOUT_WORK_DIR
+fi
+
 algorithm=( kmeans fuzzykmeans lda streamingkmeans clean)
 if [ -n "$1" ]; then
   choice=$1
@@ -59,8 +65,6 @@ fi
 echo "ok. You chose $choice and we'll use ${algorithm[$choice-1]} Clustering"
 clustertype=${algorithm[$choice-1]}
 
-WORK_DIR=/tmp/mahout-work-${USER}
-
 if [ "x$clustertype" == "xclean" ]; then
   rm -rf $WORK_DIR
   $DFSRM $WORK_DIR
@@ -98,7 +102,7 @@ if [ ! -e ${WORK_DIR}/reuters-out-seqdir
         set +e
         $DFSRM ${WORK_DIR}/reuters-sgm
         $DFSRM ${WORK_DIR}/reuters-out
-        $DFS -mkdir ${WORK_DIR}/
+        $DFS -mkdir -p ${WORK_DIR}/
         $DFS -mkdir ${WORK_DIR}/reuters-sgm
         $DFS -mkdir ${WORK_DIR}/reuters-out
         $DFS -put ${WORK_DIR}/reuters-sgm ${WORK_DIR}/reuters-sgm
diff -pruN apache-mahout-distribution-0.11.1-orig/examples/bin/cluster-syntheticcontrol.sh apache-mahout-distribution-0.11.1/examples/bin/cluster-syntheticcontrol.sh
--- apache-mahout-distribution-0.11.1-orig/examples/bin/cluster-syntheticcontrol.sh	2015-11-06 11:14:37.000000000 -0800
+++ apache-mahout-distribution-0.11.1/examples/bin/cluster-syntheticcontrol.sh	2015-11-18 14:22:51.460789000 -0800
@@ -48,7 +48,11 @@ START_PATH=`pwd`
 # Set commands for dfs
 source ${START_PATH}/set-dfs-commands.sh
 
-WORK_DIR=/tmp/mahout-work-${USER}
+if [[ -z "$MAHOUT_WORK_DIR" ]]; then
+  WORK_DIR=/tmp/mahout-work-${USER}
+else
+  WORK_DIR=$MAHOUT_WORK_DIR
+fi
 
 echo "creating work directory at ${WORK_DIR}"
 mkdir -p ${WORK_DIR}
diff -pruN apache-mahout-distribution-0.11.1-orig/examples/bin/factorize-movielens-1M.sh apache-mahout-distribution-0.11.1/examples/bin/factorize-movielens-1M.sh
--- apache-mahout-distribution-0.11.1-orig/examples/bin/factorize-movielens-1M.sh	2015-11-06 11:14:37.000000000 -0800
+++ apache-mahout-distribution-0.11.1/examples/bin/factorize-movielens-1M.sh	2015-11-18 14:22:51.483766000 -0800
@@ -43,7 +43,12 @@ fi
 export MAHOUT_LOCAL=true
 MAHOUT="$MAHOUT_HOME/bin/mahout"
 
-WORK_DIR=/tmp/mahout-work-${USER}
+if [[ -z "$MAHOUT_WORK_DIR" ]]; then
+  WORK_DIR=/tmp/mahout-work-${USER}
+else
+  WORK_DIR=$MAHOUT_WORK_DIR
+fi
+
 echo "creating work directory at ${WORK_DIR}"
 mkdir -p ${WORK_DIR}/movielens
 
@@ -77,4 +82,4 @@ shuf ${WORK_DIR}/recommendations/part-m-
 echo -e "\n\n"
 
 echo "removing work directory"
-rm -rf ${WORK_DIR}
\ No newline at end of file
+rm -rf ${WORK_DIR}
diff -pruN apache-mahout-distribution-0.11.1-orig/examples/bin/factorize-netflix.sh apache-mahout-distribution-0.11.1/examples/bin/factorize-netflix.sh
--- apache-mahout-distribution-0.11.1-orig/examples/bin/factorize-netflix.sh	2015-11-06 11:14:37.000000000 -0800
+++ apache-mahout-distribution-0.11.1/examples/bin/factorize-netflix.sh	2015-11-18 14:22:51.502751000 -0800
@@ -45,7 +45,11 @@ fi
 
 MAHOUT="../../bin/mahout"
 
-WORK_DIR=/tmp/mahout-work-${USER}
+if [[ -z "$MAHOUT_WORK_DIR" ]]; then
+  WORK_DIR=/tmp/mahout-work-${USER}
+else
+  WORK_DIR=$MAHOUT_WORK_DIR
+fi
 
 START_PATH=`pwd`
 
